### PR TITLE
Add smtp relation interface

### PR DESCRIPTION
**Pull request by arturo-seijas**
Add SMTP interface definition, as implemented by the [SMTP integrator charm](https://github.com/canonical/smtp-integrator-operator). A sample charm consuming this interface can be found [here](https://github.com/canonical/indico-operator).
This relation interface defines how the SMTP details are shared with the required charm. Note that it supports both Juju Secrets and no secrets for legacy charms.

The spec defining this interface can be found [here](https://docs.google.com/document/d/13K6TlXI2b_Hg0-0_b0LeMhPJ3WVdfJtFPsImVRgiFNc)

### CODE REVIEW

**Comment by arturo-seijas** [comment]
@PietroPasotti @simskij @gruyaume , can you please review?

File: README.md [hunk 1]
```diff
@@ -45,6 +45,7 @@ To quickly get started, see the [template interface](https://github.com/canonica
 |               | [`auth_proxy`](interfaces/auth_proxy/v0/README.md)                           |  ![Status: Draft](https://img.shields.io/badge/Status-Draft-orange)   | 
 |               | [`openfga`](interfaces/openfga/v0/README.md)                                 |  ![Status: Live](https://img.shields.io/badge/Status-Live-darkgreen)  | 
 |               | [`saml`](interfaces/saml/v0/README.md)                                       |  ![Status: Live](https://img.shields.io/badge/Status-Live-darkgreen)   |
+|               | [`smtp`](interfaces/smtp/v0/README.md)                                       |  ![Status: Live](https://img.shields.io/badge/Status-Live-darkgreen)   |
```
Comment (PietroPasotti) [inline README.md:48]: are you sure you want to go straight to live and not put it in draft first? That means you can't change the interface anymore

File: interfaces/smtp/v0/README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    Provider -- host, port, user, password, auth_type, transport_security, domain --> Requirer
+```
+
+## Behavior
+
+The requirer and the provider must adhere to a certain set of criteria to be considered compatible with the interface.
+
+### Provider
+
+- Is expected to provide the host in the relation databag.
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/README.md:34]: I usually try to use "publish" in this context to avoid confusion with the provider/requirer terminology, but that's a personal preference
```suggestion
- Is expected to publish the `host` in the relation databag.
```

File: interfaces/smtp/v0/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+### Requirer
+
+- Is not expected to provide anything
+
+## Relation Data
+
+### Provider
+
+[\[JSON Schema\]](./schemas/provider.json)
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/README.md:47]: this link is dead. Point to `[Pydantic Schema](./schema.py)` instead.

File: interfaces/smtp/v0/schema.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    Attributes:
+        NONE: none
+        NOT_PROVIDED: not_provided
+        PLAIN: plain
+    """
+
+    NONE = "none"
+    NOT_PROVIDED = "not_provided"
+    PLAIN = "plain"
+
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/schema.py:55]: newline missing. Can you `black` this whole dir?

File: interfaces/smtp/v0/schema.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        title="Port",
+        examples=[25],
+    )
+    user: Optional[str] = Field(
+        description="SMTP user.",
+        title="User",
+        examples=["some_user"],
+    )
+    password_id: Optional[str] = Field(
+        description="SMTP password.",
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/schema.py:77]:
```suggestion
        description="Juju Secret ID that resolves to a SMTP password.",
        ...
        examples=["secret:123213123123123123123"]
```

File: interfaces/smtp/v0/schema.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        title="Password",
+        examples=["somepasswd"],
+    )
+    password_id: Optional[str] = Field(
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/schema.py:81]: duplicated field?

File: interfaces/smtp/v0/schema.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        title="Domain",
+        examples=["domain"],
+    )
+
+
+class ProviderSchema(DataBagSchema):
+    """Provider schema for SMTP."""
+    app: SmtpProviderData
+
+class RequirerSchema(DataBagSchema):
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/schema.py:107]: I _think_ you can omit this class altogether and the collector won't complain. 
You can locally run `interface_tester discover --include smtp` to verify that the code you wrote is picked up correctly.

You should see something like:

```yaml
  - v0:
   - provider:
      - test_name_1
      - test_name_2
     - schema OK
     - charms:
       - charm-name (https://github.com/canonical/charm-url) custom_test_setup=no
   - requirer:
     - <no tests>
     - schema OK
     - <no charms>
```

**Review by PietroPasotti — CHANGES_REQUESTED** [review]
nothing a good linter won't fix for you, but we're not there yet

File: README.md (hunk 1 above)
Comment (arturo-seijas) [reply README.md:48]: I'm sure. The spec defining it has been closed and the charm implementing it is already in `stable`

File: interfaces/smtp/v0/README.md (hunk 2 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/README.md:34]: Good point. Thx

File: interfaces/smtp/v0/schema.py (hunk 7 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:107]: I'm afraid it's needed
```
Failed to load ProviderSchema from interfaces/smtp/v0/schema.py: schema not defined for role: provider.
Failed to load RequirerSchema from interfaces/smtp/v0/schema.py: schema not defined for role: requirer.

```

File: interfaces/smtp/v0/schema.py (hunk 6 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:83]: Fixed, thx

File: interfaces/smtp/v0/schema.py (hunk 5 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:79]: Fixed, thx

File: interfaces/smtp/v0/README.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
+# `smtp`
+
+## Overview
+
+This relation interface describes the expected behavior between of any charm claiming to be able to interface with a SMTP Provider and the SMTP Provider itself.
```
Comment (PietroPasotti) [inline interfaces/smtp/v0/README.md:5]: Aside from not being grammatically correct, this whole sentence is confusing.

In ingress we have: `This relation interface describes the expected behavior of any charm claiming to be able to provide or consume ingress data.`

We didn't yet quite agree on a generic formulation for what this should look like, and also, it's kind of repeated information in every README so IMHO we could also decide to drop this section altogether @simskij @gruyaume @benhoyt @tmihoc ?

For now, we could rephrase it to something more correct/informative:
```suggestion
This relation interface specification describes the expected behavior of charms integrating over the `smtp` Juju interface to provide or consume SMTP data.
```

Comment (PietroPasotti) [reply interfaces/smtp/v0/README.md]: see #127 for further discussion, dropping this for now

**Comment by gruyaume** [comment]
> @PietroPasotti @simskij @gruyaume , can you please review?

Please add a description to the PR. I.e. What am I reviewing here and why?

File: interfaces/smtp/v0/README.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+In most cases, this will be accomplished using the [smtp library](https://github.com/canonical/smtp-integrator-operator/blob/main/lib/charms/smtp_integrator/v0/smtp.py), although charm developers are free to provide alternative libraries as long as they fulfil the behavioural and schematic requirements described in this document.
+
+## Direction
+
+The `smtp` interface implements a provider/requirer pattern.
+The requirer is a charm that requires SMTP details to connect to an SMTP server, and the provider is a charm holding those details.
+
+```mermaid
+flowchart TD
+    Provider -- host, port, user, password_id, auth_type, transport_security, domain --> Requirer
```
Comment (gruyaume) [inline interfaces/smtp/v0/README.md:18]: Is the `password_id` a Juju secret id? If yes, please specify it somewhere here.

File: interfaces/smtp/v0/schema.py (hunk 5 above)
Comment (gruyaume) [reply interfaces/smtp/v0/schema.py:79]: @arturo-seijas , please resolve the conversation once the fix is actually made.  This applies at multiple places here.

File: interfaces/smtp/v0/README.md (hunk 9 above)
Comment (PietroPasotti) [reply interfaces/smtp/v0/README.md:18]: that also makes me think: how will either charm know it if the other end does or does not support secrets?

Comment (arturo-seijas) [reply interfaces/smtp/v0/README.md:18]: The description of the different fields is in the schema definition. I don't see how that fits here

File: interfaces/smtp/v0/schema.py (hunk 5 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:79]: The fit has been pushed. That's why I resolved. Do you see anything weird?

File: interfaces/smtp/v0/README.md (hunk 9 above)
Comment (gruyaume) [reply interfaces/smtp/v0/README.md:18]: Schema definition is also fine but even there it doesn't specifically say "Juju" secret.

Comment (arturo-seijas) [reply interfaces/smtp/v0/README.md:18]: Updated in the schema definition

**Review by gruyaume — CHANGES_REQUESTED** [review]
(as mentioned before)
Please add a description to the PR.

**Comment by benhoyt** [comment]
@arturo-seijas I'm going to be a grumpy old man for a second and double down on what @gruyaume said: you're requesting our review time (maybe 20-30 minutes each), so if you could please put in a decent amount of effort on the PR description to set context, that would be much appreciated. Simply saying "Add SMTP interface definition" doesn't add anything significant over the PR title. It may only need to be two or three sentences. Something like:

> IS DevOps has developed the `xyz` charm that provides an SMTP server. This relation interface is useful for charms connecting to such servers, providing host/port/auth details for the requiring charm. For example, see our recent charms `abc` and `def` for example consumers. Note that the relation interface being proposed supports both Juju Secrets (preferred) and no Secrets (as we still have some legacy charms) -- see README.md for more details.

All this might "go without saying" from your point of view, but for reviewers being pulled in, context and links like that are very helpful.

Also worth considering when tagging people for review is to say why you're asking them. It's somewhat obvious why you're Charm Tech (@canonical/charm-tech), but perhaps "@PietroPasotti could you please review the Scenario tests specifically? And @gruyaume we'd love your take as I know the Telco team has used SMTP in the past." (or whatever reason)

File: interfaces/smtp/v0/schema.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+class SmtpProviderData(BaseModel):
+    host: str = Field(
+        ...,
+        min_length=1,
+        description="SMTP host.",
+        title="Host",
+        examples=["example.smtp"],
+    )
+    port: int = Field(
+        None,
```
Comment (benhoyt) [inline interfaces/smtp/v0/schema.py:66]: What does the `None` here do?

File: interfaces/smtp/v0/schema.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        ge=1,
+        le=65536,
+        description="SMTP port.",
+        title="Port",
+        examples=[25],
+    )
+    user: Optional[str] = Field(
```
Comment (benhoyt) [inline interfaces/smtp/v0/schema.py:73]: Are user/password required if auth_type = plain, or is there some way to have more complex constraints like that?

File: interfaces/smtp/v0/schema.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        description="SMTP user.",
+        title="User",
+        examples=["some_user"],
+    )
+    password: Optional[str] = Field(
+        description="SMTP password.",
+        title="Password",
+        examples=["somepasswd"],
+    )
+    password_id: Optional[str] = Field(
```
Comment (benhoyt) [inline interfaces/smtp/v0/schema.py:83]: Similar with password_id -- is there a way to say it's required if using secrets, or is that known purely dynamically? If not, perhaps we can at least state the requirements in the description field.

File: interfaces/smtp/v0/schema.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+    TLS = "tls"
+
+
+class AuthType(str, Enum):
+    """Represent the auth type values.
+
+    Attributes:
+        NONE: none
+        NOT_PROVIDED: not_provided
+        PLAIN: plain
```
[the remark this answers is not in the document]
Comment (tonyandrewmeyer) [reply interfaces/smtp/v0/schema.py:49]: I presume these are SMTP AUTH methods. It seems odd that there's only PLAIN and not (e.g.) CRAM-MD5, XOAUTH2 or other common methods. Can a charm that provides the smtp interface provide additional auth methods and still be compliant, or would that require altering the specification here first?

What about providers that can handle more than one type? e.g. they can do CRAM-MD5 and prefer that over PLAIN but can do PLAIN if necessary? Is there any way to represent that?

I assume NONE means that mail will be accepted without doing any SMTP AUTH.

What is the usage of NOT_PROVIDED?

File: interfaces/smtp/v0/README.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+```
+
+or alternatively, if secrets are not supported by both sides
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/README.md:21]: Shouldn't this be "not supported by either side"?

File: interfaces/smtp/v0/schema.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        title="Host",
+        examples=["example.smtp"],
+    )
+    port: int = Field(
+        None,
+        ge=1,
+        le=65536,
+        description="SMTP port.",
+        title="Port",
+        examples=[25],
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:71]: Shouldn't the examples at least include the submission port (587), and *perhaps* 465 as well?

File: interfaces/smtp/v0/schema.py [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        title="Auth type",
+        examples=[AuthType.NONE],
+    )
+    transport_security: TransportSecurity = Field(
+        description="The security protocol to use for the outgoing SMTP relay.",
+        title="Transport security",
+        examples=[TransportSecurity.NONE],
+    )
+    domain: Optional[str] = Field(
+        description="The domain used by the sent emails from SMTP relay.",
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:99]: 'domain' is an overloaded term in email. I assume here this is the enveloper sender domain? (Sometimes called "MAIL FROM" domain). If so, could that be more explicit? Or is this meant to be either envelope sender or envelope recipient depending on whether the smtp provider is accepting mail for ~local delivery or accepting mail to send out to the world? I assume it's not any header domain, though? Or maybe the DKIM/DMARC domain? Whatever it is, it would help if this was clearer.

File: interfaces/smtp/v0/schema.py [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+        title="Domain",
+        examples=["domain"],
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:101]: This isn't a very good example of a domain :). I suggest `example.com` or anything else from RFC 2606.

File: interfaces/smtp/v0/schema.py [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+        title="Password ID",
+        examples=["secret:123213123123123123123"],
+    )
+    auth_type: AuthType = Field(
+        description="The type used to authenticate with the SMTP relay.",
+        title="Auth type",
+        examples=[AuthType.NONE],
+    )
+    transport_security: TransportSecurity = Field(
+        description="The security protocol to use for the outgoing SMTP relay.",
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:94]: This specifically says "outgoing SMTP relay" but all the others are "SMTP relay" (implying that it could be incoming or outgoing mail). Is that deliberate? If so, why?

File: interfaces/smtp/v0/schema.py [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+    PLAIN = "plain"
+
+
+class SmtpProviderData(BaseModel):
+    host: str = Field(
+        ...,
+        min_length=1,
+        description="SMTP host.",
+        title="Host",
+        examples=["example.smtp"],
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:63]: Fwiw, it would generally be better to use a RFC 2606 domain, in case one day there is a .smtp tld.

File: interfaces/smtp/v0/README.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+```yaml
+related-units: {}
+application_data: {
+  "host": "example.smtp",
+  "port": "25",
+  "user": "example_user",
+  "password_id": "01548499c9233d4612352c989162d940f6a9e6f6d5cc058dfcf66f51575e09c2",
+  "auth_type": "plain",
+  "transport_security": "tls",
+  "domain": "domain",
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/README.md:62]: It would be nice to have a better example than `domain` here.

File: interfaces/smtp/v0/README.md (hunk 9 above)
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/README.md:18]: People have differing opinions about whether a username is secret or not. Is there a Canonical standard here? I think I have seen elsewhere that in a case like this the Juju secret would have all the credential, ie. both username and password, and the databag would only have the secret ID.

**Review by tonyandrewmeyer — COMMENTED** [review]
A few comments that tbh are more from someone long in the email world than specifically charm-tech comments :)

File: docs/json_schemas/smtp/v0/provider.json [hunk 21]
```diff
@@ -0,0 +1,129 @@
+{
```
[the remark this answers is not in the document]
Comment (simskij) [reply docs/json_schemas/smtp/v0/provider.json:1]: That's correct.

File: interfaces/smtp/v0/schema.py [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+from typing import Optional
+
+
+class TransportSecurity(str, Enum):
+    """Represent the transport security values.
+
+    Attributes:
+        NONE: none
+        STARTTLS: starttls
+        TLS: tls
```
[the remark this answers is not in the document]
Comment (simskij) [reply interfaces/smtp/v0/schema.py:35]: Either remove it or provide a useful description of what each option means.

File: interfaces/smtp/v0/schema.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (simskij) [reply interfaces/smtp/v0/schema.py:49]: Same as above: make the comment useful or remove it.

File: interfaces/smtp/v0/README.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+## Behavior
+
+The requirer and the provider must adhere to a certain set of criteria to be considered compatible with the interface.
+
+### Provider
+
+- Is expected to publish the `host` in the relation databag.
+- Is expected to publish the `port` in the relation databag.
+- Is expected to publish the `auth_type` in the relation databag.
+- Is expected to publish the `transport_security` in the relation databag.
```
Comment (simskij) [inline interfaces/smtp/v0/README.md:37]: These expectations aren't really useful as they're all implicitly provided through the schema. This section is supposed to describe the behavior, rather than the schema itself. See https://github.com/canonical/charm-relation-interfaces/blob/main/interfaces/grafana_auth/v0/README.md#behavior for an example.

**Review by simskij — CHANGES_REQUESTED** [review]

File: interfaces/smtp/v0/README.md (hunk 9 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/README.md:18]: Not that I'm aware of

File: interfaces/smtp/v0/schema.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py]: Removed

File: interfaces/smtp/v0/schema.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py]: Removed

File: interfaces/smtp/v0/README.md (hunk 23 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/README.md]: Noted. Thx

File: interfaces/smtp/v0/schema.py (hunk 11 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:61]: I could state that in the field description, if you think that'd be useful. I don't think pydantic enables us to define those kind of constraints

File: interfaces/smtp/v0/schema.py (hunk 16 above)
Comment (arturo-seijas) [reply interfaces/smtp/v0/schema.py:87]: Yes, that's correct. By explicit, you mean renaming the field or does expanding the description suffice?

Comment (tonyandrewmeyer) [reply interfaces/smtp/v0/schema.py:85]: I think expanding the description would be fine.

File: =4
Comment (tonyandrewmeyer) [inline =4:1]: Accidental add/commit?

File: interfaces/smtp/v0/schema.py [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+        title="Auth type",
+        examples=[AuthType.NONE],
+    )
+    transport_security: TransportSecurity = Field(
+        description="The security protocol to use for the SMTP relay.",
+        title="Transport security",
+        examples=[TransportSecurity.NONE],
+    )
+    domain: Optional[str] = Field(
+        description="The EMAIL FROM domain for the outgoing email.",
```
Comment (tonyandrewmeyer) [inline interfaces/smtp/v0/schema.py:85]:
```suggestion
        description="The MAIL FROM domain for the outgoing email.",
```

**Review by tonyandrewmeyer — APPROVED** [review]
Looks great! Thanks for making those adjustments. One tiny nit and one I assume accidental file add.

File: =4
Comment (arturo-seijas) [reply =4:1]: Oops, yes. Thx

**Review by benhoyt — APPROVED** [review]
Looks good to me now (that definitely get @simskij's final review). And thanks for the update to the PR description.

**Review by simskij — APPROVED** [review]

**Review by gruyaume — APPROVED** [review]
